### ISSUE TITLE
multiple stream merge step

### ISSUE THREAD

**Issue by brendanheywood** [catalyst/moodle-tool_dataflows#120]
This is very distinct from the various 'join' type steps. A join step takes data from multiple streams and depending on the data in them combines them into a single record in the output iterator.

This merge step on the other hand simple interleaves the incoming records together. The key config and logic here is that it can merge streams which were originally split apart from one stream back together and it can do it while retaining the original order.

A logical inverse of this step is the case flow step in #119 

The way this will work is that the engine will record in the transient iterator data, scoped to each step, a start and stop time for each record as it flows through. So for any give record we will have a complete timeline of all the previous steps in the flow.

The join step will need to be configured with an expression which is how to interleave the records, which is a simple comparison. By default if it detects that all of the streams that it is pulling from have a common ancestor then it will choose the finish time of that closest common ancestors processing time as the comparison. If there is no ancestor and there is no expression then it will not be valid.

By necessity it will need a small buffer of a single record for each iterator it pulls from, and once it can compare all the records values it can choose the earliest value and then keep pulling again.

There is an important edge case to work out. Lets say we have a stream of odd random digits, and we split that using the flow case step into odd and even digits. The odd output stream will be identical to the starting stream. The even digits stream will be empty. The merge step can't blindly keep asking the even digit stream for more values, because that will cascade up the chain until it gets to the case which splits them and it will iterate through its entire input looking for even values and then get to the end and close the iterator. For small iterators this will work just fine, but for anything large this will result in the entire iterator being loaded into an in memory buffer at the case steps output for the odd output stream, which will only be pulled once the merge gets the signal that the evens are empty.

So we need to figure out same way to that an iterator is blocked temporarily, I think flow steps new an extra blocking state:

https://github.com/catalyst/moodle-tool_dataflows/issues/53#issue-1232695929

**Comment by keevan** [comment]
Potentially, the in memory buffer, once it reaches a certain size, can (internal to the relevant step) store its _state_ to the scratch space (disk) and keep appending to it (somewhat similar but not exactly the same as disk swapping). This moves the memory concerns to disk which shouldn't be a concern (space is typically cheaper). Once it needs to flow through, indicated by finally getting an even digit, or when the iterator has reached the end (no more numbers), then it will read and push the remaining odd digits to flow through?

**Comment by brendanheywood** [comment]
Yeah and it is probably a premature optimization, even if you start with an sql reader step the postgres db driver actually loads this into memory anyway in pages of 100k records:

https://github.com/moodle/moodle/blob/master/lib/dml/pgsql_native_moodle_database.php#L61

So we really need to hit some large scale before this bites us and then we can tackle either moving the buffer, or implementing some sort of flow regulation feedback mechanism.

A few different steps will need buffers so that will be a shared object and it can handle it all internally

**Comment by brendanheywood** [comment]
Also even if we pick a more realistic example but still with highly skewed data, so lets say 95% of records are odd and 5% are even, even if we have a simple round robin pulling from both iterators we will on average end up with a buffer on one side with 1 record and the buffer on the other side with only 19 records. So I'm very comfortable kicking this down the road until its a real problem.
